nav2_localization/src/nav2_localization.cpp

```diff
@@ -0,0 +1,254 @@
+
+    initial_pose_sub_ = create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+        "initialpose", rclcpp::SystemDefaultsQoS(),
+        std::bind(&LocalizationServer::initialPoseReceived, this, std::placeholders::_1));
+}
+
+void
+LocalizationServer::initPlugins()
+{
+    auto node = shared_from_this();
+
+    try {
+        sample_motion_model_type_ = nav2_util::get_plugin_type_param(node, sample_motion_model_id_);
+        sample_motion_model_ = sample_motion_model_loader_.createUniqueInstance(sample_motion_model_type_);
+    } catch (const pluginlib::PluginlibException & ex) {
+        RCLCPP_FATAL(get_logger(), "Failed to create sample motion model. Exception: %s", ex.what());
+        exit(-1);
+    }
+
+    sample_motion_model_->configure(node);
+
+    try {
+        matcher2d_type_ = nav2_util::get_plugin_type_param(node, matcher2d_id_);
+        matcher2d_ = matcher2d_loader_.createUniqueInstance(sample_motion_model_type_);
+    } catch (const pluginlib::PluginlibException & ex) {
+        RCLCPP_FATAL(get_logger(), "Failed to create matcher2d. Exception: %s", ex.what());
+        exit(-1);
+    }
+
+    matcher2d_->configure(node, map_);
+
+    try {
+        solver_type_ = nav2_util::get_plugin_type_param(node, solver_id_);
+        solver_ = solver_loader_.createUniqueInstance(solver_type_);
+    } catch (const pluginlib::PluginlibException & ex) {
+        RCLCPP_FATAL(get_logger(), "Failed to create solver. Exception: %s", ex.what());
+        exit(-1);
+    }
+
+    solver_->configure(node, sample_motion_model_, matcher2d_, nav_msgs::msg::Odometry(), geometry_msgs::msg::Pose());
+}
+
+void
+LocalizationServer::initialPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
+{
+    // TODO
+}
+
+void
+LocalizationServer::laserReceived(sensor_msgs::msg::LaserScan::ConstSharedPtr laser_scan)
+{
+    // Since the sensor data is continually being published by the simulator or robot,
+    // we don't want our callbacks to fire until we're in the active state
+    if (!get_current_state().id() == lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE) {return;}
+
+    geometry_msgs::msg::TransformStamped odom_to_base_transform;
+    try
+    {
+        odom_to_base_transform = tf_buffer_->lookupTransform(odom_frame_id_,
+                                    base_frame_id_,
+                                    laser_scan->header.stamp,
+                                    transform_tolerance_);
+    }
+    catch(const tf2::TransformException& e)
+    {
+        RCLCPP_ERROR(get_logger(), "%s", e.what());
+        return;
+    }
+
+    // Change the stamped tf into an odom msg for the solver to use
+    nav_msgs::msg::Odometry current_odom = transformStampedToOdom(odom_to_base_transform);
+    // The estimated robot's pose in the global frame
+    geometry_msgs::msg::PoseWithCovariance current_pose = solver_->solve(current_odom, *laser_scan);
```

Comment:
This is where you should be using all the plugins you have access to. So if the matcher needs the scan, give it the scan. Then the matcher already has the most recent info when the solver just wants to "solve?" it (I'm still not really clear on the "solver" name). 

So if we have the solver, matcher, and motion models, I think at this point we give the appropriate inputs to each module to reduce the hardcoded logic in the "solver" which ideally we want as little application code in as possible. Ideally, it would be only the MCL and its interactions with those other modules. So anything that will be necessary for all sets of plugins, we should try to do here (like giving sensor data, odometry updates, etc to the other modules so that it doesn't get done in the solver).